### ISSUE TITLE
[Bug] Out-of-order messages with +2 generations corrupts sender ratchet until epoch update/reset

### ISSUE THREAD

**Issue by mefsantos** [openmls/openmls#453]
**Describe the bug**

The issue occurs when messages to be processed are out-of-order with a generation gap >= 2, i.e., a sender sends __at least__ three messages (generation gap >= 2), and the receiver processes the latter first. 

**To Reproduce**

1. Alice creates the group; 
2. Alice invites Bob; 
3. Bob joins the group (new_from_welcome)
4. Alice processes the added member commit

5. Alice attemps to send 3 consecutive messages:
  `let mls_msg_1 = try alice_group.create_message(b"message1")`
  `let mls_msg_2 = try alice_group.create_message(b"message2")`
  `let mls_msg_3 = try alice_group.create_message(b"message3")`

6. (for some reason) bob receives and processes the third message first:
 `let bob_event = try bob_group.process_messages([mls_msg_3])`

7. the output event will be `invalidCiphertext` caused by  a decryption error on `aead_open` in `MLSCiphertext.to_plaintext`

**Expected behavior**

Bob was expected to be able to decrypt the out-of-order message since the generation gap is still within `OUT_OF_ORDER_TOLERANCE` bounds.

**Actual behavior**

Bob was not able to decrypt the message because he is not correctly deriving the generation secret used to derive the ratchet secret. Concretely, the bug is caused by always using the base generation (`self.generation`) and not the actual "current" generation in which the ratchet secret is being derived in `ratchet_secret()` ([here](https://github.com/openmls/openmls/blob/30842a9768f1f46e620d1b61f9a11e416192d249/openmls/src/tree/sender_ratchet.rs#L96)).

**Platform (please complete the following information):** 
its a code-based bug, not platform-dependent. However, just for the record:
 - OS & OS version: android 10 and iOS 14
 - Architecture: arm64-v8a and arm64 
 - Version: Java and Swift ports from the branch [main-a454a5a](https://github.com/openmls/openmls/commit/a454a5a649f5663ce299c29dedf54938240e68ef). However the "bugged" code is the same in the latest commit

**Suggested Solutions**
1. [here](https://github.com/openmls/openmls/blob/30842a9768f1f46e620d1b61f9a11e416192d249/openmls/src/tree/sender_ratchet.rs#L53-L60) move the loop range into a local variable and increment `self.generation` inside the loop, after deriving the `past_secret`, like so:
```rust
let gap = generation - self.generation; 
for _ in (0..gap) {
    if self.past_secrets.len() == OUT_OF_ORDER_TOLERANCE as usize {
        self.past_secrets.remove(0);
    }
    let new_secret =
        self.ratchet_secret(ciphersuite, self.past_secrets.last().unwrap());
    self.past_secrets.push(new_secret);
    self.generation += 1; // <- bugfix
}
```
2. [here](https://github.com/openmls/openmls/blob/30842a9768f1f46e620d1b61f9a11e416192d249/openmls/src/tree/sender_ratchet.rs#L91-L98) pass the current generation as argument in the function `ratchet_secret()` like so:
```rust

 // ...
          for gen_gap in 0..(generation - self.generation) { // loop range index (gen_gap) is now required
              if self.past_secrets.len() == OUT_OF_ORDER_TOLERANCE as usize {
                  self.past_secrets.remove(0);
              }
              let new_secret =
                  self.ratchet_secret(ciphersuite, self.past_secrets.last().unwrap(), (self.generation + gen_gap);
              self.past_secrets.push(new_secret);
          }

// ...

fn ratchet_secret(&self, ciphersuite: &Ciphersuite, secret: &Secret, generation: u32) -> Secret {
        derive_tree_secret(
            secret,
            "secret",
            NodeIndex::from(self.index).as_u32(),
            generation, // self.generation replaced by function parameter
            ciphersuite.hash_length(),
        )
    }
```

This way, the `new_secret` used to derive the ratchet secrets is derived on the correct generation.

**Edits**
1. In solution 1, since it's a _ForEach_, the loop higher range doesn't _need_ to be set outside of the `for` declaration since the difference will only be computed once. So the actual bugFix is just one line, the `self.generation` increment 
2. rolled back Edit 1_ (added the local variable with the generation gap) as suggested by @raphaelrobert

**Comment by kkohbrok** [comment]
Hi @mefsantos and thanks for the detailed bug report! This is certainly an area of the code that hasn't seen a lot of testing yet and needs some love for other reasons as well (#243, #103 and #18 come to mind, although they're somewhat overlapping). The test vectors we have been working on lately are also primarily testing group evolution and not so much the actual message encryption.

Your first solution seems like the easiest fix for now, so I'd vote for that given that we will probably have to touch a lot of the code here anyway. @franziskuskiefer @raphaelrobert any objections?

**Comment by raphaelrobert** [comment]
Thanks for flagging it @mefsantos! I agree with @kkohbrok that option 1 is probably easiest for now. I would however recommend to first introduce the gap variable you propose – just to be on the safe side: `gap = generation - self.generation` and use that as the upper bound for the loop. Also, [line 63](https://github.com/openmls/openmls/blob/30842a9768f1f46e620d1b61f9a11e416192d249/openmls/src/tree/sender_ratchet.rs#L63) is then probably redundant.

Lastly, we clearly need tests for all of this stuff.

Do you feel like submitting a PR for that?

**Comment by mefsantos** [comment]
I've spoken with my team (since I am not a RUST developer) and __we__ will submit the PR. Is it enough to patch the bug and implement one unit test for this purpose (out-of-order messages) in the PR?

**Comment by raphaelrobert** [comment]
Yes, that's the idea. The PR should address the problem, have test coverage (with at least positive and a negative test) and make the CI happy.

**Comment by raphaelrobert** [comment]
This was fixed in #456.
